### PR TITLE
Refactor staging delegate code

### PR DESCRIPTION
**Pull request by chanseokoh**
Slight refactoring. Stagers can know which project to deploy at the instantiation time, so we don't need to pass a project when calling their `stage()` method.

This fixes the problem of https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2416#issuecomment-331533357 (not the issue itself).

### CODE REVIEW

**Comment by chanseokoh** [comment]
Need to fix a test.

**Comment by chanseokoh** [comment]
Test fixed. Ready for review.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/StagingDelegate.java [hunk 1]
```diff
@@ -29,20 +29,25 @@
...
    * @param stderrOutputStream where staging may stream stderr
    */
-  IStatus stage(IProject project, IPath stagingDirectory, IPath safeWorkDirectory,
+  IStatus stage(IPath stagingDirectory, IPath safeWorkDirectory,
       MessageConsoleStream stdoutOutputStream, MessageConsoleStream stderrOutputStream,
       IProgressMonitor monitor);
 
   /**
+   * Returns a {#link ISchedulingRule} for this stager to work safely. For example, if this stager
+   * needs to lock down a {#link IProject} during staging, it could return the {#link IProject}.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/StagingDelegate.java]: nit: a --> an

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/flex/FlexMavenPackagedProjectStagingDelegate.java [hunk 2]
```diff
@@ -156,4 +161,10 @@ protected IPath getDeployArtifact(IProject project, IPath safeWorkingDirectory,
     }
   }
 
+  @Override
+  public ISchedulingRule getSchedulingRule() {
+    Preconditions.checkNotNull(project);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/flex/FlexMavenPackagedProjectStagingDelegate.java]: Why isn't this null check in the constructor? The field is final.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/flex/FlexWarStagingDelegate.java [hunk 3]
```diff
@@ -35,15 +37,24 @@
...
+  protected IPath getDeployArtifact(IPath safeWorkDirectory, IProgressMonitor monitor)
+      throws CoreException {
     IPath war = safeWorkDirectory.append("app-to-deploy.war");
     WarPublisher.publishWar(project, war, monitor);
     return war;
   }
+
+  @Override
+  public ISchedulingRule getSchedulingRule() {
+    Preconditions.checkNotNull(project);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/flex/FlexWarStagingDelegate.java]: ditto. Why not put the null check in the constructor?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/StagingDelegate.java [hunk 4]
```diff
@@ -29,20 +29,25 @@
...
    * @param stdoutOutputStream where staging may stream stdout
    * @param stderrOutputStream where staging may stream stderr
    */
-  IStatus stage(IProject project, IPath stagingDirectory, IPath safeWorkDirectory,
+  IStatus stage(IPath stagingDirectory, IPath safeWorkDirectory,
       MessageConsoleStream stdoutOutputStream, MessageConsoleStream stderrOutputStream,
       IProgressMonitor monitor);
 
   /**
+   * Returns a {#link ISchedulingRule} for this stager to work safely. For example, if this stager
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/StagingDelegate.java]: nit: a --> an

**Pull request merged**